src/dmd/dcast.d

```diff
@@ -2565,6 +2565,28 @@ Expression castTo(Expression e, Scope* sc, Type t, Type att = null)
         }
     }
 
+    // Casting to noreturn isn't an actual cast
+    // Rewrite cast(<qual> noreturn) <exp>
+    // as      <exp>, assert(false)
+    if (t.isTypeNoreturn())
+    {
+        // Only generate assert(false) if e won't abort
+        if (!e.type.isTypeNoreturn())
+        {
+            auto ini = t.defaultInitLiteral(e.loc);
+            return Expression.combine(e, ini);
+        }
+
+        // Paint a copy in case of different type qualifiers
+        if (t != e.type)
+        {
+            e = e.copy();
```

Comment:
Why is this copy needed? Can't we simply set `e`s type directly?

Reply:
Thought it was required to not affect the initial expression.
But the tests in `noreturn3.d` still work without the `copy`, hence I've removed the copy